### ISSUE TITLE
[49.19-Snapshot] NPE When Attempting to Save (Manual or Auto)

### ISSUE THREAD

**Issue by IllianiCBT** [MegaMek/mekhq#3978]
### Environment
Origin Project: MekHQ
Java Vendor: Eclipse Adoptium
Java Version: 11.0.22
Platform: Mac OS X 13.6.6 (x86_64)

### Description
NPE encountered when attempting manual save. Reloading to a prior save resolved the issue.
<img width="545" alt="image" src="https://github.com/MegaMek/mekhq/assets/103902653/dcd31bba-072b-4674-9157-365fb3294638">

### Log
```00:24:26,619 INFO  [mekhq.gui.CampaignGUI] {AWT-EventQueue-0}
mekhq.gui.CampaignGUI.saveCampaign(CampaignGUI.java:1309) - Saving campaign...

00:24:37,023 ERROR [mekhq.gui.CampaignGUI] {AWT-EventQueue-0}
mekhq.gui.CampaignGUI.saveCampaign(CampaignGUI.java:1353) - 
java.lang.NullPointerException
	at megamek.common.EntityListFile.writeCrewAttributes(EntityListFile.java:1239)
	at megamek.common.EntityListFile.writeEntityList(EntityListFile.java:832)
	at mekhq.campaign.mission.BotForce.writeToXML(BotForce.java:329)
	at mekhq.campaign.mission.Scenario.writeToXMLBegin(Scenario.java:784)
	at mekhq.campaign.mission.Scenario.writeToXML(Scenario.java:755)
	at mekhq.campaign.mission.Mission.writeToXMLBegin(Mission.java:220)
	at mekhq.campaign.mission.Contract.writeToXMLBegin(Contract.java:661)
	at mekhq.campaign.mission.AtBContract.writeToXMLBegin(AtBContract.java:770)
	at mekhq.campaign.mission.Mission.writeToXML(Mission.java:202)
	at mekhq.campaign.Campaign.writeToXML(Campaign.java:4154)
	at mekhq.gui.CampaignGUI.saveCampaign(CampaignGUI.java:1344)
	at mekhq.gui.CampaignGUI.saveCampaign(CampaignGUI.java:1317)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.AbstractButton.doClick(AbstractButton.java:369)
	at java.desktop/com.apple.laf.ScreenMenuItem.actionPerformed(ScreenMenuItem.java:129)
	at java.desktop/java.awt.MenuItem.processActionEvent(MenuItem.java:690)
	at java.desktop/java.awt.MenuItem.processEvent(MenuItem.java:649)
	at java.desktop/java.awt.MenuComponent.dispatchEventImpl(MenuComponent.java:375)
	at java.desktop/java.awt.MenuComponent.dispatchEvent(MenuComponent.java:363)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:775)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

**Comment by IllianiCBT** [comment]
Looks like autosaves are also affected by this bug.

```18:18:57,537 ERROR [mekhq.service.AutosaveService] {AWT-EventQueue-0}
mekhq.service.AutosaveService.performAutosave(AutosaveService.java:96) - 
java.lang.NullPointerException
	at megamek.common.EntityListFile.writeCrewAttributes(EntityListFile.java:1239)
	at megamek.common.EntityListFile.writeEntityList(EntityListFile.java:832)
	at mekhq.campaign.mission.BotForce.writeToXML(BotForce.java:329)
	at mekhq.campaign.mission.Scenario.writeToXMLBegin(Scenario.java:784)
	at mekhq.campaign.mission.Scenario.writeToXML(Scenario.java:755)
	at mekhq.campaign.mission.Mission.writeToXMLBegin(Mission.java:220)
	at mekhq.campaign.mission.Contract.writeToXMLBegin(Contract.java:661)
	at mekhq.campaign.mission.AtBContract.writeToXMLBegin(AtBContract.java:770)
	at mekhq.campaign.mission.Mission.writeToXML(Mission.java:202)
	at mekhq.campaign.Campaign.writeToXML(Campaign.java:4154)
	at mekhq.service.AutosaveService.performAutosave(AutosaveService.java:89)
	at mekhq.service.AutosaveService.requestDayAdvanceAutosave(AutosaveService.java:59)
	at mekhq.campaign.Campaign.newDay(Campaign.java:3456)
	at mekhq.campaign.CampaignController.advanceDay(CampaignController.java:82)
	at mekhq.gui.CampaignGUI.lambda$initTopButtons$59(CampaignGUI.java:1067)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279)
	at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:298)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6635)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6400)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5011)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)```
	
Attempting to load the save gets me this...
```18:22:40,815 ERROR [mekhq.campaign.io.CampaignXmlParser] {SwingWorker-pool-2-thread-1}
mekhq.campaign.io.CampaignXmlParser.parse(CampaignXmlParser.java:107) - 
org.xml.sax.SAXParseException; lineNumber: 4115; columnNumber: 377; XML document structures must start and end within the same entity.
	at java.xml/com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:261)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:339)
	at java.xml/javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:122)
	at mekhq.campaign.io.CampaignXmlParser.parse(CampaignXmlParser.java:105)
	at mekhq.campaign.CampaignFactory.createCampaign(CampaignFactory.java:83)
	at mekhq.gui.dialog.DataLoadingDialog$Task.doInBackground(DataLoadingDialog.java:336)
	at mekhq.gui.dialog.DataLoadingDialog$Task.doInBackground(DataLoadingDialog.java:192)
	at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

18:22:40,852 ERROR [mekhq.gui.dialog.DataLoadingDialog.Task] {AWT-EventQueue-0}
mekhq.gui.dialog.DataLoadingDialog$Task.done(DataLoadingDialog.java:365) - 
java.util.concurrent.ExecutionException: mekhq.campaign.io.CampaignXmlParseException: org.xml.sax.SAXParseException; lineNumber: 4115; columnNumber: 377; XML document structures must start and end within the same entity.
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at java.desktop/javax.swing.SwingWorker.get(SwingWorker.java:613)
	at mekhq.gui.dialog.DataLoadingDialog$Task.done(DataLoadingDialog.java:361)
	at java.desktop/javax.swing.SwingWorker$5.run(SwingWorker.java:750)
	at java.desktop/javax.swing.SwingWorker$DoSubmitAccumulativeRunnable.run(SwingWorker.java:847)
	at java.desktop/sun.swing.AccumulativeRunnable.run(AccumulativeRunnable.java:112)
	at java.desktop/javax.swing.SwingWorker$DoSubmitAccumulativeRunnable.actionPerformed(SwingWorker.java:857)
	at java.desktop/javax.swing.Timer.fireActionPerformed(Timer.java:317)
	at java.desktop/javax.swing.Timer$DoPostEvent.run(Timer.java:249)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: mekhq.campaign.io.CampaignXmlParseException: org.xml.sax.SAXParseException; lineNumber: 4115; columnNumber: 377; XML document structures must start and end within the same entity.
	at mekhq.campaign.io.CampaignXmlParser.parse(CampaignXmlParser.java:108)
	at mekhq.campaign.CampaignFactory.createCampaign(CampaignFactory.java:83)
	at mekhq.gui.dialog.DataLoadingDialog$Task.doInBackground(DataLoadingDialog.java:336)
	at mekhq.gui.dialog.DataLoadingDialog$Task.doInBackground(DataLoadingDialog.java:192)
	at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.xml.sax.SAXParseException; lineNumber: 4115; columnNumber: 377; XML document structures must start and end within the same entity.
	at java.xml/com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:261)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:339)
	at java.xml/javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:122)
	at mekhq.campaign.io.CampaignXmlParser.parse(CampaignXmlParser.java:105)
	... 9 more```

I've been unable to determine the state of events that leads to the bug, but I was able to isolate it to the below save. My guess is that something isn't saving correctly, causing the file to become malformed.

[IllianiCBT_StateWithBug.cpnx.gz](https://github.com/MegaMek/mekhq/files/14899346/IllianiCBT_StateWithBug.cpnx.gz)
[IllianiCBT_StateBeforeBug.cpnx.gz](https://github.com/MegaMek/mekhq/files/14899347/IllianiCBT_StateBeforeBug.cpnx.gz)

**Comment by IllianiCBT** [comment]
Extracting and viewing the save, attached above, I can see that one of the lines seems to end abruptly and without a closing brace:

```
<entity chassis="Goliath" model="GOL-1H" type="Quad" commander="false" offboard="false" hidden="false" deployment="2" deploymentZone="-1" deploymentZoneWidth="3" deploymentZoneOffset="0" deploymentZoneAnyNWx="-1" deploymentZoneAnyNWy="-1" deploymentZoneAnySEx="-1" deploymentZoneAnySEy="-1" neverDeployed="true" externalId="174ac0e3-7bd9-4201-a602-846a396ec48c">
			<pilot size="1" name="Ionel Stanisic" nick="" gender="MALE" clanperson="false" gunnery="3" piloting="3" externalId="f28c45ea-baaf-40cd-9662-5531d1590793" ejected="false" advantages="dodge_maneuver::tm_forest_ranger::tm_mountaineer::sniper::sensor_geek" edge="edge 2::edge_when_headhit::edge_when_tac::edge_when_ko::edge_when_explosion::edge_when_masc_fails" autoeject="true
```

**Comment by Thom293** [comment]
That quoted code is one of the SC allied Ace modifiers.  But unless someone else changed it, it has been unchanged since it was introduced.  Ive drawn that one at least twice and it worked fine...

Regardless, i got this bug too.  Previous Autosave and Log attached.  
[mekhq.log](https://github.com/MegaMek/mekhq/files/14901878/mekhq.log)

[Autosave-2-Crazed Profanity-30700206.cpnx.gz](https://github.com/MegaMek/mekhq/files/14901906/Autosave-2-Crazed.Profanity-30700206.cpnx.gz)
[Customs.zip](https://github.com/MegaMek/mekhq/files/14901915/Customs.zip)

**Comment by Thom293** [comment]
Ok, not certain, and it will take me forever to go back and check old units for a contract where I got base attacks, but I think that previously, Base Attacks could not get any modifierss.  It was simply the OpFor and your allies, but it could not roll enemy reinforcements or Air Support or any other modifier.  There are no modifiers listed here, but it is certainly rolling extra enemy forces for some reason. 

Maybe wrong, but that is my recollection.  So seems like something has changed.
![image](https://github.com/MegaMek/mekhq/assets/23645569/34b95c2b-dec0-48be-b855-74c399ad7329)

**Comment by AaronGullickson** [comment]
The issue is this line in EntityListFile:

```java
if (entity.game.getOptions().booleanOption(OptionsConstants.RPG_CONDITIONAL_EJECTION)) {
```

The immediate reason for the error is that BotForce entities have null games, and so this breaks. Looking through EntityListFile, there are several places already where `game != null` is checked so this appears to be a case that was missed. 

The larger reason this is happening is PR #3906 where I switched BotForce from writing entities with MHQXMLUtility to writing with EntityListFile. The main reason I did this is because EntityListFile has done a better job of tracking new additional fields than MHQXMLUtility. MHQXMLUtility doesn't even track conditional auto-ejection at all! However, EntityListFile also is not necessarily aware that it might be getting entities without games. The secondary reason for this is that having two different XML writers is nuts and we need to work to consolidate them so this seemed the place to start. So, I don't think reverting that change is the right approach.

I think there are two solutions to the problem. The short term solution would be to add checks on game to this line like it is for other lines. Also, I should check to see if there are other places that game is not checked. the longer term solution would be to avoid this more systematically by passing game options directly into the writer methods.
